File: solutions/1759. Count Number of Homogenous Substrings/1759.rs
// Time complexity: O(n)
// Space complexity: O(1)
impl Solution {
    pub fn count_homogenous(s: String) -> i32 {
        const K_MOD: i32 = 1_000_000_007;
        let mut ans = 0;
        let mut count = 0;
        let mut current_char = '@';

        for c in s.chars() {
            count = if c == current_char { count + 1 } else { 1 };
            current_char = c;
            ans += count;
            ans %= K_MOD;
        }

        ans
    }
}

File: solutions/1759. Count Number of Homogenous Substrings/1759.cpp
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
 public:
  int countHomogenous(string_view s) {
    constexpr int kMod = 1'000'000'007;
    int ans = 0;
    int count = 0;
    char currentChar = '@';

    for (const char c : s) {
      count = c == currentChar ? count + 1 : 1;
      currentChar = c;
      ans += count;
      ans %= kMod;
    }

    return ans;
  }
};